File: 18/code_from_slides/timer.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <chrono>

// Create shorter aliases
using time_point = std::chrono::steady_clock::time_point;
constexpr auto now = &std::chrono::steady_clock::now;

double elapsed_time(time_point start, 
                    time_point finish) {
    return (finish - start).count() / 1000000000.0;
}

int main() {
    std::cout << "Enter some event names, and I'll time them!\n" 
              << "Press Control-d to exit" << std::endl;

    std::vector<time_point> times;
    std::vector<std::string> events;
    
    times.push_back(now());
    events.push_back("Begin");
    
    std::string line;
    while(std::cin) {
        std::getline(std::cin, line);
        times.push_back(now());
        events.push_back(line);
    }
    
    std::cout << std::setw(20) << "Event Description" 
              << std::setw(15) << "From Start" 
              << std::setw(15) << "From Previous" << std::endl;
    
    std::cout << std::fixed << std::setprecision(3);
    for(int i=1; i<times.size(); ++i) {   
        std::cout << std::setw(20) << events[i] 
                  << std::setw(15) << elapsed_time(times[0],   times[i]) 
                  << std::setw(15) << elapsed_time(times[i-1], times[i]) 
                  << " seconds" << std::endl;
    }
}

File: 18/code_from_slides/Timer.java
import java.util.ArrayList;
import java.util.Scanner;

import java.time.Instant;
import java.time.Duration;

class Timer {
    private static double elapsedTime(Instant start, Instant finish) {
        return ((double) Duration.between(start, finish).toMillis()) / 1000.0;
    }
    public static void main(String[] args) {
        System.out.println("Enter some event names, and I'll time them!"); 
        System.out.println("Press Control-d to exit");
    
        Scanner in = new Scanner(System.in);
        String line = "";
        
        ArrayList<Instant> times = new ArrayList<>();
        ArrayList<String> events = new ArrayList<>();
        times.add(Instant.now());
        events.add("Begin");
        
        while(in.hasNextLine()) {
            line = in.nextLine();
            times.add(Instant.now());
            events.add((line != null) ? line : "End");
        }
        
        System.out.printf("%20s %15s %15s\n", 
                           "Event Description", "From Start", "From Previous");
        for(int i=1; i<times.size(); ++i) {
            System.out.printf("%20s %15.3f %15.3f seconds\n", events.get(i),
                elapsedTime(times.get(0),   times.get(i)),
                elapsedTime(times.get(i-1), times.get(i)));
        }   
    }
}
